spring-kafka/src/main/java/org/springframework/kafka/listener/DeadLetterPublishingRecoverer.java

```diff
@@ -659,59 +660,69 @@ private void enhanceHeaders(Headers kafkaHeaders, ConsumerRecord<?, ?> record, E
+						() -> consumerGroup.getBytes(StandardCharsets.UTF_8), HeaderNames.HeadersToAdd.GROUP);
 			}
 		}
 	}
 
-	private void maybeAddHeader(Headers kafkaHeaders, String header, byte[] value, HeaderNames.HeadersToAdd hta) {
+	private void maybeAddHeader(Headers kafkaHeaders, String header, Supplier<byte[]> valueSupplier,
+			HeaderNames.HeadersToAdd hta) {
+
 		if (this.whichHeaders.contains(hta)
 				&& (this.appendOriginalHeaders || kafkaHeaders.lastHeader(header) == null)) {
-			kafkaHeaders.add(header, value);
+			kafkaHeaders.add(header, valueSupplier.get());
 		}
 	}
 
 	private void addExceptionInfoHeaders(Headers kafkaHeaders, Exception exception, boolean isKey,
 			HeaderNames names) {
 
-		appendOrReplace(kafkaHeaders, new RecordHeader(isKey ? names.exceptionInfo.keyExceptionFqcn
-				: names.exceptionInfo.exceptionFqcn,
-				exception.getClass().getName().getBytes(StandardCharsets.UTF_8)), HeaderNames.HeadersToAdd.EXCEPTION);
+		appendOrReplace(kafkaHeaders,
+				() -> new RecordHeader(
+						isKey ? names.exceptionInfo.keyExceptionFqcn : names.exceptionInfo.exceptionFqcn,
+						exception.getClass().getName().getBytes(StandardCharsets.UTF_8)),
+				HeaderNames.HeadersToAdd.EXCEPTION);
 		if (exception.getCause() != null) {
-			appendOrReplace(kafkaHeaders, new RecordHeader(names.exceptionInfo.exceptionCauseFqcn,
-					exception.getCause().getClass().getName().getBytes(StandardCharsets.UTF_8)),
+			appendOrReplace(kafkaHeaders,
+					() -> new RecordHeader(
+							names.exceptionInfo.exceptionCauseFqcn,
+							exception.getCause().getClass().getName().getBytes(StandardCharsets.UTF_8)),
 					HeaderNames.HeadersToAdd.EX_CAUSE);
 		}
 		String message = exception.getMessage();
 		if (message != null) {
-			appendOrReplace(kafkaHeaders, new RecordHeader(isKey
-					? names.exceptionInfo.keyExceptionMessage
-					: names.exceptionInfo.exceptionMessage,
-					exception.getMessage().getBytes(StandardCharsets.UTF_8)), HeaderNames.HeadersToAdd.EX_MSG);
+			appendOrReplace(kafkaHeaders,
+					() -> new RecordHeader(
+							isKey ? names.exceptionInfo.keyExceptionMessage : names.exceptionInfo.exceptionMessage,
+							exception.getMessage().getBytes(StandardCharsets.UTF_8)),
+					HeaderNames.HeadersToAdd.EX_MSG);
 		}
-		appendOrReplace(kafkaHeaders, new RecordHeader(isKey
-				? names.exceptionInfo.keyExceptionStacktrace
-				: names.exceptionInfo.exceptionStacktrace,
-				getStackTraceAsString(exception).getBytes(StandardCharsets.UTF_8)),
+		appendOrReplace(kafkaHeaders,
+				() -> new RecordHeader(
+						isKey ? names.exceptionInfo.keyExceptionStacktrace : names.exceptionInfo.exceptionStacktrace,
+						getStackTraceAsString(exception).getBytes(StandardCharsets.UTF_8)),
 				HeaderNames.HeadersToAdd.EX_STACKTRACE);
 	}
 
-	private void appendOrReplace(Headers headers, RecordHeader header, HeaderNames.HeadersToAdd hta) {
+	private void appendOrReplace(Headers headers, Supplier<RecordHeader> headerSupplier, HeaderNames.HeadersToAdd hta) {
```

Comment:
Just as a cosmetic suggestion that might clean up the code a little bit, I wonder if we could change this method's signature to use `String header, Supplier<byte[]> valueSupplier` instead of `Supplier<RecordHeader> header supplier`, and then just call the `headers.add()` method with those like we do at `maybeAddHeaders`.

Of course, it doesn't really change anything, but since we're adding a few elements it might help keeping the overall size down, and also would make those method's signatures more consistent.

Reply:
Yes, it indeed looks neater with the proposed change.